### ISSUE TITLE
Pickle error in Parallel

### ISSUE THREAD

**Issue by dennisrpnw** [joblib/joblib#706]
Hey,

I get an error with 0.12 that didn't occur in 0.11

snippet of code:
`num_cores = multiprocessing.cpu_count()`
`chunks = _get_chunks(ids, 100)`
`logger.info('cores: {0}, chunks: {1}'.format(num_cores, len(chunks)))`
`Parallel(n_jobs=num_cores)(delayed(_clean_text)(c) for c in chunks)`

The resulting error is:
`Cannot pickle files that are not opened for reading: a`

**Comment by tomMoral** [comment]
Thanks for reporting this issue.

Could you give us more details so we can reproduce this locally? In particular, could you include the full traceback of the `Error` in order for us to see which lines generated this issue and the code of the `_clean_text` function?

Note that in general, it is important to give Minimal Complete Verifiable Examples when you repport issues, as described in this [blog post](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

**Comment by lesteve** [comment]
I completely agree with @tomMoral about providing a MCVE, this is really key to get good quality feed-back! A full stack-trace is super useful too!

Taking a wild guess after googling the error, it seems like:
* the error happens in cloudpickle
* cloudpickle can only pickle files opened in read mode and you are using `mode='a'`

I am slightly surprised that it worked with joblib 0.11 because it was using the stdlib pickle and pickle can not pickle files I think. I may be missing something though.

**Comment by gVkWY8NJAa** [comment]
This may or may not be related, if not I apologize, but this error appears in 0.12 and not in 0.11

The below error occurs when executing "In [7]:" section of this notebook: https://github.com/CamDavidsonPilon/Probabilistic-Programming-and-Bayesian-Methods-for-Hackers/blob/master/Chapter1_Introduction/Ch1_Introduction_PyMC3.ipynb which is
```
### Mysterious code to be explained in Chapter 3.
with model:
    step = pm.Metropolis()
    trace = pm.sample(10000, tune=5000,step=step)
```

Downgrading to 0.11 resolves this issue

```
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/foo/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 333, in _sendback_result
    exception=exception))
  File "/home/foo/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/backend/queues.py", line 233, in put
    obj = CustomizableLokyPickler.dumps(obj, reducers=self._reducers)
  File "/home/foo/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/backend/reduction.py", line 145, in dumps
    p.dump(obj)
TypeError: can't pickle CVM objects
"""

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<ipython-input-10-2bd7fdec376c> in <module>()
      2 with model:
      3     step = pm.Metropolis()
----> 4     trace = pm.sample(10000, tune=5000,step=step)

~/miniconda3/lib/python3.6/site-packages/pymc3/sampling.py in sample(draws, step, init, n_init, start, trace, chain_idx, chains, cores, tune, nuts_kwargs, step_kwargs, progressbar, model, random_seed, live_plot, discard_tuned_samples, live_plot_kwargs, compute_convergence_checks, use_mmap, **kwargs)
    440         _print_step_hierarchy(step)
    441         try:
--> 442             trace = _mp_sample(**sample_args)
    443         except pickle.PickleError:
    444             _log.warning("Could not pickle model, sampling singlethreaded.")

~/miniconda3/lib/python3.6/site-packages/pymc3/sampling.py in _mp_sample(**kwargs)
    980         traces = Parallel(n_jobs=cores)(jobs)
    981     else:
--> 982         traces = Parallel(n_jobs=cores, mmap_mode=None)(jobs)
    983 
    984     return MultiTrace(traces)

~/miniconda3/lib/python3.6/site-packages/joblib/parallel.py in __call__(self, iterable)
    960 
    961             with self._backend.retrieval_context():
--> 962                 self.retrieve()
    963             # Make sure that we get a last message telling us we are done
    964             elapsed_time = time.time() - self._start_time

~/miniconda3/lib/python3.6/site-packages/joblib/parallel.py in retrieve(self)
    863             try:
    864                 if getattr(self._backend, 'supports_timeout', False):
--> 865                     self._output.extend(job.get(timeout=self.timeout))
    866                 else:
    867                     self._output.extend(job.get())

~/miniconda3/lib/python3.6/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    513         AsyncResults.get from multiprocessing."""
    514         try:
--> 515             return future.result(timeout=timeout)
    516         except LokyTimeoutError:
    517             raise TimeoutError()

~/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/_base.py in result(self, timeout)
    429                 raise CancelledError()
    430             elif self._state == FINISHED:
--> 431                 return self.__get_result()
    432             else:
    433                 raise TimeoutError()

~/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/_base.py in __get_result(self)
    380     def __get_result(self):
    381         if self._exception:
--> 382             raise self._exception
    383         else:
    384             return self._result

TypeError: can't pickle CVM objects
```

**Comment by lesteve** [comment]
> This may or may not be related, if not I apologize, but this error appears in 0.12 and not in 0.11

General guideline: if you are unsure, open a separate issue with a stand-alone snippet to reproduce and link to the other issue that may be related in the description.

I can reproduce using the notebook (please next time, do your best to create a stand-alone snippet, it makes it a more lot likely to get good quality feedback).

It seems like the problem comes from the fact that  pymc3 catches PickleError to fall-back to single threaded mode:

```pytb
~/miniconda3/lib/python3.6/site-packages/pymc3/sampling.py in sample(draws, step, init, n_init, start, trace, chain_idx, chains, cores, tune, nuts_kwargs, step_kwargs, progressbar, model, random_seed, live_plot, discard_tuned_samples, live_plot_kwargs, compute_convergence_checks, use_mmap, **kwargs)
    440         _print_step_hierarchy(step)
    441         try:
--> 442             trace = _mp_sample(**sample_args)
    443         except pickle.PickleError:
    444             _log.warning("Could not pickle model, sampling singlethreaded.")```

But now that joblib uses cloudpickle the error you get is a `TypeError`. Not 100% sure what's the best way to fix that, maybe catches all errors produced by cloudpickle.dump and reraise a PickleError?

**Comment by lesteve** [comment]
@dennisrpnw if you provide a full traceback, we can figure out whether the two issues have the same cause.

**Comment by lesteve** [comment]
Here is a stand-alone snippet for the pymc3-related issue:

```py
import numpy as np

import pymc3 as pm

rng = np.random.RandomState(0)
count_data = rng.randint(1, 70, 100).astype(dtype=np.float64)
n_count_data = len(count_data)

with pm.Model() as model:
    alpha = 1.0/count_data.mean()  # Recall count_data is the
                                   # variable that holds our txt counts
    lambda_1 = pm.Exponential("lambda_1", alpha)
    lambda_2 = pm.Exponential("lambda_2", alpha)
    
    tau = pm.DiscreteUniform("tau", lower=0, upper=n_count_data - 1)

with model:
    step = pm.Metropolis()
    trace = pm.sample(5, tune=10,step=step)
```

**Comment by dennisrpnw** [comment]
Sorry, I was in the middle of a deployment on our production environment.

Traceback:
```
2018-06-26 10:43:52 - ERROR - error cleaning, process, error: Cannot pickle files that are not opened for reading: a
Traceback (most recent call last):
  File "clean.py", line 111, in _process
    Parallel(n_jobs=num_cores)(delayed(_clean_text)(c) for c in chunks)
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/joblib/parallel.py", line 962, in __call__
    self.retrieve()
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/joblib/parallel.py", line 865, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/joblib/_parallel_backends.py", line 515, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/joblib/externals/loky/_base.py", line 431, in result
    return self.__get_result()
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/joblib/externals/loky/_base.py", line 382, in __get_result
    raise self._exception
PicklingError: Cannot pickle files that are not opened for reading: a
```

Code of _clean_text()
```
def _clean_text(pids):
    global db_engine
    ct = ClearText()
    csession = scoped_session(sessionmaker(bind=db_engine))
    for id in pids:
        try:
            ft = csession.query(PackageFulltext).filter(PackageFulltext.package_id == id).first()
            text_clear, wrong = ct.clear_text(ft.text)
            ft.text_clear = text_clear
        except Exception, e:
            logger.error('error cleaning, package id: {0}, error: {1}'.format(id, str(e)))
    csession.commit()
    csession.close()
    csession.remove()
```

ct.clear_text() just removes spaces when there is more than one between two words and some other string operations.

**Comment by lesteve** [comment]
@dennisrpnw thanks for posting additional information unfortunately that does not help very much. I understand it's some effort to put together a minimal reproducible example but without it we are a bit in the dark ... Something that would require less effort from you and may help would be to be able to look at the code of clean.py.

Thinking about it a bit more, it does not seem related to @gVkWY8NJAa's problem (which I moved to #707) since the exception is a PicklingError so it's not related to `Parallel` changing its exception type when something goes wrong (probably under the hood because joblib use cloudpickle instead of pickle).

**Comment by lesteve** [comment]
@tomMoral @ogrisel just for explicitness' sake I don't intend on working on fixing either on these two problems in the near future.

**Comment by ogrisel** [comment]
@lesteve @dennisrpnw it's very likely that there is an open file description in the `global db_engine` instance.

I that case it can work with old multiprocessing-based backend with the default fork context but would have probably crashed under windows.

In any way, relying on a closed at fork file descriptor in a python object is probably a bad idea. Better refactor that code to avoid relying on such a global instance with an open file descriptor in it. We won't add pickle support for open file objects in 'append' mode in cloudpickle any time soon.

**Comment by EloyRoura** [comment]
Hello,

Sorry to do not solve the issue but keep asking on that. I think I have a different problem although still with pickle. I am just testing out the joblib library with that:

````
import multiprocessing
from joblib import Parallel, delayed
def some_funtion(p):
    return p*p

parameters = range(10)
num_cores=4#multiprocessing.cpu_count()
Parallel(n_jobs=num_cores, verbose=1)(delayed(some_funtion)(p) for p in parameters
````

and this is the error with its Traceback:
```
[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   3 out of  10 | elapsed:    0.0s remaining:    0.1s
joblib.externals.loky.process_executor._RemoteTraceback: 

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/joblib/externals/loky/backend/queues.py", line 151, in _feed
    obj, reducers=reducers)
  File "/usr/local/lib/python3.5/dist-packages/joblib/externals/loky/backend/reduction.py", line 145, in dumps
    p.dump(obj)
  File "/usr/local/lib/python3.5/dist-packages/joblib/parallel.py", line 258, in __getstate__
    for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/dist-packages/joblib/parallel.py", line 258, in <listcomp>
    for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/dist-packages/joblib/parallel.py", line 231, in _wrap_non_picklable_objects
    need_wrap = "__main__" in getattr(obj, "__module__", "")
TypeError: argument of type 'NoneType' is not iterable
```

any help with that would be great!

Eloy

**Comment by EloyRoura** [comment]
Sorry, my bad I just needed some more patience... nothing to do with your library and may be out of this scope. Just in case, I'd like to post the answer here below.

Basically, I figured out the call needs to go inside the main, as follows:

```
import multiprocessing
from joblib import Parallel, delayed
from multiprocessing import Pool

#EM example in https://homes.cs.washington.edu/~jmschr/lectures/Parallel_Processing_in_Python.html

def some_funtion(p):
    # print(p*p)
    return p*p

def processInput(i):
    return i * i

if __name__ == '__main__':
    # With joblib
    parameters = range(10)
    print(Parallel(n_jobs=num_cores, verbose=1)(delayed(some_funtion)(p) for p in parameters))

    #With built in
    num_cores = multiprocessing.cpu_count()
    with Pool(num_cores) as p:
        print(p.map(processInput, range(10)))
```

**Comment by ogrisel** [comment]
@EloyRoura the point of using the loky backend in joblib is to replace the multiprocessing.Pool backend. With the loky backend you no longer need to protect your code with the `if __name__ == '__main__'` condition.


Here is an example:

```
>>> from joblib import Parallel, delayed
>>> def some_funtion(p):
...     return p*p
... 
>>> parameters = range(10)
>>> num_cores=4#multiprocessing.cpu_count()
>>> Parallel(n_jobs=num_cores, verbose=1)(delayed(some_funtion)(p) for p in parameters)
[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  10 out of  10 | elapsed:    0.4s finished
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
```
